observatory-dags/observatory/dags/telescopes/jstor.py

```diff
@@ -0,0 +1,505 @@
+
+    def download(self):
+        """ Downloads Google Books reports.
+
+        :return: the paths on the system of the downloaded files.
+        """
+        for report in self.reports_info:
+            url = report['url']
+            download_path = self.download_path(report['type'])
+            success = download_report(url, download_path)
+
+    def transform(self):
+        """ Transform a Jstor release into json lines format and gzip the result.
+
+        :return: None.
+        """
+        for file in self.download_files:
+            results = []
+            with open(file) as tsv_file:
+                csv_reader = csv.DictReader(tsv_file, delimiter='\t')
+                for row in csv_reader:
+                    transformed_row = OrderedDict((convert(k), v) for k, v in row.items())
+                    results.append(transformed_row)
+
+            report_type = 'country' if 'country' in file else 'institution'
+            list_to_jsonl_gz(self.transform_path(report_type), results)
+
+    def cleanup(self) -> None:
+        """ Delete files of downloaded, extracted and transformed release. Add to parent method cleanup and assign a
+        label to the gmail messages that have been processed.
+
+        :return: None.
+        """
+        super().cleanup()
+
+        service = create_gmail_service()
+        label_id = get_label_id(service, JstorTelescope.PROCESSED_LABEL_NAME)
+        for report in self.reports_info:
+            message_id = report['id']
+            body = {
+                'addLabelIds': [label_id]
+            }
+            response = service.users().messages().modify(userId='me', id=message_id, body=body).execute()
+            try:
+                message_id = response['id']
+                logging.info(f"Added label '{JstorTelescope.PROCESSED_LABEL_NAME}' to GMAIL message, message_id: "
+                             f"{message_id}")
+            except KeyError:
+                raise AirflowException(f'Unsuccessful adding label to GMAIL message, message_id: {message_id}')
+
+
+class JstorTelescope(SnapshotTelescope):
+    """
+    The JSTOR telescope.
+
+    Saved to the BigQuery tables: <project_id>.jstor.jstor_countryYYYYMMDD and
+    <project_id>.jstor.jstor_institutionYYYYMMDD
+    """
+
+    DAG_ID_PREFIX = 'jstor'
+    PROCESSED_LABEL_NAME = 'processed_report'
+
+    # download settings
+    MAX_ATTEMPTS = 3
+    FIXED_WAIT = 20  # seconds
+    MAX_WAIT_TIME = 60 * 10  # seconds
+    EXP_BASE = 3
+    MULTIPLIER = 10
+
+    def __init__(self, organisation: Organisation, extra: dict, dag_id: Optional[str] = None,
+                 start_date: datetime = datetime(2015, 9, 1), schedule_interval: str = '@monthly',
+                 dataset_id: str = 'jstor', source_format: str = SourceFormat.NEWLINE_DELIMITED_JSON,
+                 dataset_description: str = '', catchup: bool = False, airflow_vars: List = None,
+                 airflow_conns: List = None):
```

Comment:
I think it would be clearer to specify `publisher_id` as a parameter instead of `extra` in the constructor.